Improve backsub input validation and testing

jmuhlich opened the pull request:
Rework of #82. Original features retained:
* Make `remove` column non-required.
* Don't allow backsub to run if no channels are linked to their respective background channels in the markersheet.
* Implement correct validation behavior for backsub-related markersheet columns. Error messages include marker_name or other identifying information to help the user find the problem more easily.
* Drop rows marked with `remove` from markersheet before generating input markersheet for quantification so that markersheet rows match up with image channels.
* Add tests for backsub input validation checks.
* Fix backsub test case in workflow test so it actually triggers backsub properly.

Updates relative to the original PR:
* Simplify markersheet handling by marking all fields in the schema as `meta` with a default value of `null`. Update all relevant places in the code and tests to use these maps instead of the old lists.
* Streamline error checking code branching structure in backsub input validation.
* Harmonize error messages with common nf-core practice and improved internal consistency.
* Port markersheet validation tests from workflow to function tests for easier input management (inputs can then be defined inline in the test script rather than requiring a discrete csv file).

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Make sure your code lints (`nf-core pipelines lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).

jmuhlich committed:
Utilize nf-schema meta key to simplify sheet handling

jmuhlich committed:
Fix backsub test input to actually subtract something

Co-authored-by: Krešimir Beštak <kbestak@gmail.com>

jmuhlich committed:
Improve backsub markersheet column validation

Co-authored-by: Krešimir Beštak <kbestak@gmail.com>

jmuhlich committed:
Drop removed rows from markersheet for quantification

Co-authored-by: Krešimir Beštak <kbestak@gmail.com>

jmuhlich committed:
Add validation tests for backsub marker metadata

Co-authored-by: Krešimir Beštak <kbestak@gmail.com>

jmuhlich committed:
Improve backsub validation logic flow a little

kbestak commented on subworkflows/local/utils_nfcore_mcmicro_pipeline/main.nf:
```diff
@@ -192,54 +184,56 @@ def validateInputMarkersheet( markersheet_data ) {
-            error('background column values must exist in the marker_name column. The following background column values do not exist in the marker_name column: ' + outliers_list)
+    // Validate backsub data
+    if (!params.backsub) {
+        def backsub_columns = ['exposure', 'background', 'remove']
+        if (markersheet_data*.subMap(backsub_columns)*.collect{ c, v -> v != null }.flatten().any()) {
```
If the remove column is provided, params.backsub is false, if just a warning is passed, MCQUANT would fail due to channel removal logic. Adding an error targeting this case 
```suggestion
        def backsub_columns = ['exposure', 'background']
        if (markersheet_data*.subMap('remove')*.collect{ c, v -> v != null }.flatten().any()) {
            error("The remove column is present in the marker sheet, but params.backsub is set to false. Please set params.backsub to true or remove TRUE values from the remove column.")
        }
        else if (markersheet_data*.subMap(backsub_columns)*.collect{ c, v -> v != null }.flatten().any()) {
```

kbestak commented on subworkflows/local/utils_nfcore_mcmicro_pipeline/tests/validate.nf.test:
```diff
@@ -0,0 +1,111 @@
+            //assert function.stdout*.contains("Subtraction will NOT be performed").any()
+        }
+    }
+
```
If adding error for params.backsub = false and remove values added
```suggestion

    test("Error if backsub false but remove is provided") {
        function "validateInputMarkersheet"
        when {
            params {
                backsub = true
            }
            function {
                """
                input[0] = [
                    [marker_name:"M1", cycle_number:1, channel_number:1, remove: true],
                ]
                """
            }
        }
        then {
            assert function.failed
            assert function.stdout*.contains("The remove column is present in the marker sheet").any()
        }
    }

```

kbestak commented on subworkflows/local/utils_nfcore_mcmicro_pipeline/tests/validate.nf.test:
```diff
@@ -0,0 +1,111 @@
+                """
+                input[0] = [
+                    [marker_name:"M1", cycle_number:1, channel_number:1, exposure: 100],
+                    [marker_name:"M2", cycle_number:1, channel_number:2, remove: true],
```
If adding error for params.backsub = false and remove values added since this would now fail istead of warn
```suggestion
                    [marker_name:"M2", cycle_number:1, channel_number:2, background:"M1"],
```

kbestak reviewed (approved):
Hey @jmuhlich, thank you for making time to add the changes. This is a very nice and elegant approach that allows the markersheet to be expanded in the future in a very straightforward and logical manner.

I've only noticed that if remove values are given and backsub is set to false, a warning is given, but the pipeline fails on MCQUANT due to the remove marker logic on the nextflow side. My suggestion would be to throw an error for this case as you can see above.

Looks good to me! :rocket:

jmuhlich committed:
Fix quantification markersheet generation bug

When backsub is not being used, we now properly ignore the "remove" channel
meta property.

jmuhlich commented:
Thanks for the suggested changes, but I'm afraid that solution defeats the design goal of allowing the user to enable/disable backsub without also needing to change their marker sheet. I implemented an alternative fix which corrects the quantification markersheet generation, not removing channels if params.backsub is false.
